### DISCUSSION TITLE
Issue while migrating Micronaut 3.X to 4.X

### DISCUSSION THREAD

**Discussion by bardale** [micronaut-projects/micronaut-core#9680] [Q&A]
Hello,
I have multimodule project developed using Micronaut, Kotlin and Java. 
Currently I am using below versions 
Java : (java 11) - OpenJDK 64-Bit Server VM Zulu11
Kotlin : 1.6.21
Micronaut: 3.7.4
Gradle: 7.3.3

Also below are some plugin and their corresponding versions:
id("io.micronaut.application") version "3.7.0"
id "io.micronaut.library" version "3.6.5"
id "org.jetbrains.kotlin.jvm" version "1.6.21"
id "org.jetbrains.kotlin.kapt" version "1.6.21"
id "org.jetbrains.kotlin.plugin.allopen" version "1.6.21"
id "com.github.johnrengelman.shadow" version "7.1.2"

I have updated all version as per https://micronaut.io/2023/05/09/upgrade-to-micronaut-framework-4-0-0/  to migrate to micronaut 4
So new versions will be
Java : OpenJDK 64-Bit Server VM Zulu17.42+20-SA (build 17.0.7+7-LTS, mixed mode, sharing)
Kotlin : 1.8.21
Micronaut: 4.0.0
Gradle: 8.0
and plugins to
id("io.micronaut.application") version "4.0.0"
id "io.micronaut.library" version "4.0.0"
id "org.jetbrains.kotlin.jvm" version "1.8.21"
id "org.jetbrains.kotlin.kapt" version "1.8.21"
id "org.jetbrains.kotlin.plugin.allopen" version "1.8.21"
id "com.github.johnrengelman.shadow" version "8.1.1"

Still I am getting below error while executing gradlew.bat clean build
FAILURE: Build failed with an exception.

* What went wrong:
java.lang.NoClassDefFoundError: org/jetbrains/kotlin/gradle/plugin/KotlinBasePlugin
> org/jetbrains/kotlin/gradle/plugin/KotlinBasePlugin

Can someone please help?

**Comment by graemerocher** [comment]
did you upgrade the Gradle version to 8?

**Reply by bardale** [reply]
Greetings @graemerocher 
Yes I have updated gradle version to 8

**Comment by wetted** [comment]
@bardale Are you on Gradle 8.2.1. I think that's required for the Micronaut plugins. Please try that.

**Reply by bardale** [reply]
@wetted Thanks for your reply. Even its not worked with 8.2.1. Still getting same error.

**Comment by wetted** [comment]
@bardale In that case, I don't see anything you provided so far that is suspect otherwise. If I create a default Micronaut 4 application using Micronaut Launch (i.e. `mn create-app --build=gradle --jdk=17 --lang=kotlin --test=junit com.example.demo`), the resulting project builds just fine. 

Can you provide an example project that demonstrates the build failure?

**Comment by wetted** [comment]
@bardale is this still a problem for you? Try the latest versions. You also need to upgrade Kotlin to 1.9.x (1.9.20 is latest). Micronaut 4 doesn't use Kotlin 1.8.x.

**Reply by bardale** [reply]
Hello @wetted 
Greetings.

Problem is resolved by using below versions.
- 'kotlin'  : '1.8.20'
-  'micronautSerialization': '2.2.4',
-  'coroutines'            : '1.7.3',
- used com.google.devtools.ksp version "1.8.20-1.0.11" instead of Kapt
- micronautVersion=4.1.6
- gradle-8.3

**Reply by wetted** [reply]
OK. Glad you got it resolved. You should still upgrade kotlin to 1.9.x for Micronaut 4 projects.
